### ISSUE TITLE
CUDA memory leak in multi-processing

### ISSUE THREAD

**Issue by rajcscw** [pytorch/pytorch#44156]
## 🐛 CUDA memory leak in multi-processing

I'm working on parallelizing my training tasks (that are independent) in a multi-GPU setting. For this, I have created several workers (processes) that take a list of tasks and work parallelly to finish them. Each task is assigned with a device and any process that executes it will use the assigned device.

**The problem I observe is**: once a process finishes a task, the memory that was allocated in a GPU is not released. The same worker might work on another task that uses a different GPU and once it is done, it holds that memory too. In this way, a process occupies some parts of memory in all GPUs and never releases them until the process is terminated. This leads to memory leak and eventually results in CUDA out of memory error if several worker processes are used.

**Note:** I have tried using `torch.cuda.empty_cache()` to release the memory, only unsuccessful.

**However**, if each worker process is assigned a device to work with and any task that is executed by that process would always use its designated device, then this behavior is not observed. Since a worker process always uses the same GPU for its tasks, it seems like it releases/reuses the same allocated memory.

## To Reproduce

To reproduce this behavior, let's consider two settings :

- **Device on task**: Each task is assigned a device and any tensor operations are performed on the specified device irrespective of which worker executes it **(this results in memory leak)**

```python
from torch import multiprocessing as mp
import torch
import random


class TaskWithDevice:
    def __init__(self, device: str):
        self.device = device

    def run(self):
       rand_tensor = torch.rand(1000, 1000).to(self.device)


class Worker(mp.Process):
    def __init__(self, queue: mp.Queue):
        super().__init__(target=self.work, args=(queue,))

    def work(self, queue):
        while queue.qsize() > 0:
            task = queue.get()
            task.run()
            del task


if __name__ == '__main__':
    mp.set_start_method('spawn')
    n_tasks = 10000

    # available devices
    available_devices = [f"cuda:{i}" for i in range(torch.cuda.device_count())]
    devices = random.choices(available_devices, k=n_tasks)

    # task queue
    task_queue = mp.Queue()
    for i in range(n_tasks):
        task = TaskWithDevice(devices[i])
        task_queue.put(task)

    # workers
    workers = [Worker(task_queue) for device in available_devices]
    for worker in workers:
        worker.start()
    for worker in workers:
        worker.join()

```

- **Device on worker**: Each worker is assigned a device and any tasks that are executed by it would always use the same device **(here there is no memory leak and it seems like the allocated memory is reused for tasks by that worker)**

```python
from torch import multiprocessing as mp
import torch


class Task:
    def run(self, device: str):
        rand_tensor = torch.rand(1000, 1000).to(device)


class WorkerWithDevice(mp.Process):
    def __init__(self, queue: mp.Queue, device: str):
        self.device = device
        super().__init__(target=self.work, args=(queue,))

    def work(self, queue):
        while queue.qsize() > 0:
            task = queue.get()
            task.run(self.device)
            del task


if __name__ == '__main__':
    mp.set_start_method('spawn')
    n_tasks = 10000

    # available devices
    available_devices = [f"cuda:{i}" for i in range(torch.cuda.device_count())]

    # task queue
    task_queue = mp.Queue()
    for i in range(10000):
        task = Task()
        task_queue.put(task)

    # workers attached to a particular device
    workers = [WorkerWithDevice(task_queue, device) for device in available_devices]
    for worker in workers:
        worker.start()
    for worker in workers:
        worker.join()

```

## Expected behavior

In the first setting, it is expected that once the task is finished, the GPU memory should be released as soon as the program finishes `run()`

## Environment

PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.2 LTS (x86_64)
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB

Nvidia driver version: 450.51.06
cuDNN version: Could not collect

**Comment by SsnL** [comment]
You mean `empty_cache` after `task.run()` in `Worker.run` doesn't work?

**Comment by le1nux** [comment]
I came across the same issue when using the multiprocessing pool. Similar to @rajcscw 's example, I assigned a list of tasks to GPUs and then ran them across a pool of processes. This resulted in GPU memory overflows as well. 

One workaround is to recreate the process after each job completion, as shown by this snippet:
```python
def work(job: Job):
    # run task
    pass

job_list: List[Job] = # init your job list, where each Job has been assigned to a GPU

with Pool(processes=30, maxtasksperchild=1) as pool:
    pool.map(func=work, iterable=job_list, chunksize=1)
```

We initialize a pool that recreates the process after each **task**, specified by `maxtasksperchild = 1`. The pool implementation defines a task as a single **chunk**, where each chunk in return is a list of Jobs to be run. By setting `maxtasksperchild=1`in `Pool` and  `chunksize = 1` in `map()`, we therefore ensure that after each job completion, the process is in fact recreated and does not accumulate GPU memory on various GPUs.

**Comment by rajcscw** [comment]
Sure? Could you please check with the example I provided for **Device on task**? If you check the memory usage and compare with the example for **Device on worker**, you would clearly see that the total memory usage (summed up across all GPUs) is higher indicating there is a memory leak, although both use the same number of workers and they are supposed to use the same amount of memory at any point of time.

**Comment by SsnL** [comment]
@rajcscw Yeah I was using your device on task example, not my code. Adding `empty_cache` shows no "leak" really, other than the cuda contexts.

**Comment by le1nux** [comment]
@SsnL This is interesting. Did you run it on multiple GPUs? 

In the mulitprocessing pool, I have basically the following setup:
```
----------------------------------time-axis------------------>
process 1 || job 1 @ GPU1 | job 5 @ GPU 4 | job 9 @ GPU 2  | 
process 2 || job 2 @ GPU2 | job 6 @ GPU 3 | job 10 @ GPU 1 |
process 3 || job 3 @ GPU3 | job 7 @ GPU 2 | job 11 @ GPU 4 |
process 4 || job 4 @ GPU4 | job 8 @ GPU 1 | job 12 @ GPU 3 |
```
 What I saw was that `process 1` would allocate some memory on `GPU 1` to execute `job 1`. After `job 1`  finishes, even calling `empty_cache` would leave a high amount of memory allocated on `GPU 1`. Thus, in the next round `process 4` would allocate even more memory on `GPU 1` for `job 8` and also not releasing most of it. And so finally, when `process 2` runs `job 10` on `GPU 1` so little memory would be available that we get an overflow. 

When I was running models that allocated multiple GB of GPU memory also multiple GB would not be released. I'm certainly no cuda expert, but I assume the cuda context should not be of that size? What helped though, was just killing the process, cause thereby the GPU memory was automatically released.

**Comment by SsnL** [comment]
@le1nux I just ran @rajcscw 's 1st script with `empty_cache` on a 4 GPU machine. Each context is usually of size 300~800MB. Generally you should make sure to clean up any references to GPU tensors (including indirect ones e.g., a tensor backed by a computation graph which contain many intermediate buffers) before calling that.

**Comment by oryondark** [comment]
No, `empty_cache` is not suitable in multi processing using python.
If you run with multi threading on worker, python left parameters( no gradients when you using torch no grad ). 
This means that per thread may must conflict on memory before empty cache. So we have to consider batch size how to handle our large scale data, and/or discuss with your team for Model compression ( Pruning or Huffman )

In my example, i did reducing tensor batch size and append one line for empty cache below inference codes.

**Comment by kopyl** [comment]
So it's not fixed?
